text/0000-bang-type.md

```diff
@@ -0,0 +1,401 @@
+never contain an element, a `!` enum variant makes the variant guaranteed never
+to occur and so forth. It might seem pointless to use a `!` function argument
+or a `Vec<!>` (just as it would be pointless to use a `()` function argument or
+a `Vec<()>`), but that's no reason to disallow it. And generic code sometimes
+requires it.
+
+Rust already has empty types in the form of empty enums. Any code that could be
+written with this RFC's `!` can already be written by swapping out `!` with
+`Never` (sans implicit casts, see below). So if this RFC could create any
+issues for the language (such as making it unsound or complicating the
+compiler) then these issues would already exist for `Never`.
+
+# Detailed design
+
+Add a type `!` to Rust. `!` behaves like an empty enum except that it can be
+implicitly cast to any other type. ie. the following code is acceptable:
+
+```rust
+let r: Result<i32, !> = Ok(23);
+let i = match r {
+    Ok(i)   => i,
+    Err(e)  => e, // e is cast to i32
+}
+```
+
+Implicit casting is necessary for backwards-compatibility so that code like the
+following will continue to compile:
+
+```rust
+let i: i32 = match some_bool {
+    true  => 23,
+    false => panic!("aaah!"), // an expression of type `!`, gets cast to `i32`
+}
+
+match break {
+    ()  => 23,  // matching with a `()` forces the match argument to be cast to type `()`
+}
+```
+
+In the compiler, remove the distinctions that treat diverging and converging
+expressions as two different kinds of things (eg. stuff like `FnConverging` vs
+`FnDiverging`). Use the type system to do things like reachability analysis.
+
+Add an implementation for `!` of any trait that it can trivially implement. Add
+methods to `Result<T, !>` and `Result<!, E>` for safely extracting the inner
+value. Name these methods along the lines of `unwrap_nopanic`, `safe_unwrap` or
+something.
+
+# Drawbacks
+
+Someone would have to implement this.
+
+# Alternatives
+
+  * Don't do this.
+  * Move @reem's `Void` type into `libcore`. This would create a standard empty
+    type and make it available for use in the standard libraries. If we were to
+    do this it might be an idea to rename `Void` to something else (`Never`,
+    `Empty` and `Mu` have all been suggested). Although `Void` has some
+    precedence in languages like Haskell and Idris the name is likely to trip
+    up people coming from a C/Java et al. background as `Void` is *not* `void`
+    but it can be easy to confuse the two.
+
+# Unresolved questions
+
+Apparently, rust used to have something similar to this but it was removed.
+There are still a few references to `ty_bot` in the compiler. Why was this
+taken out? Note that if there any arguments for not having type `!` in the
+language they should apply equally well to `Never`/`Void` so I assume the old
+`ty_bot` was trying to be something crazier than this RFC's `!` (such as a
+subtype of all types, given the name). Could someone who was around back then
+clarify this?
```

Comment:
@eddyb 

> Resolution is not "propagated", inference variables are being created and updated during the traversal of a function, and then the types of all the nodes and side-table miscellanea are resolved (which queries the state of the inference variables where necessary).

Ah, so quasi-unification in a side table as the traverse happens.

> I don't understand why `<rvalue> as _` is necessary, when just creating a new diverging inference variable would have the same effect, if diverging inference variables are always defaulted to ! instead of ().

There is two things going on here. The first is limiting whatever magic we do to rvalues, which I think is good idea no matter what:
```rust
struct Baz;
struct Foo { bar: Baz }
fn asdf() {
    let x: ! = panic();
    x.bar = Baz; // WTF
}
```

The second is adding this coercion under the hood. I agree that is not necessary, and basically an implementation strategy that shouldn't be noticeable. But I suspect it might actually help with the implementation: as opposed to have seemingly normal expressions be type checked rather oddly, we associated all the magic with one AST/IR node (the conversion).

Also, I'd immediately like to immediately deprecate the coercion once ! becomes a real type. Perhaps if `panic!` and friends can be changed to be `for<T> fn() -> T`, there ergonomic fallout will be acceptable. Pruning the no-op `! -> !` conversions from the AST/IR, and then warning on the others seems like the easiest way to do this.